### PR TITLE
Add a predicated wait for Lock::ConditionVariable

### PR DESCRIPTION
**Pull request by Leont**
## The Problem
Currently, the `wait` method of `Lock::ConditionVariable` follows the paradigm of the C functions. I would suggest adding a more rakuish variant that takes a predicate. Much like the `protect` method in`Lock`, this makes it much easier to do the correct thing (and people will do things incorrectly if the easiest way is not also the correct way). So I propose adding this:

```
multi method wait(&predicate) {
    self.wait until predicate;
}
```

To take an example from `Awaiter`, instead of writing:
```
loop {
    $l.protect: {
        last if $remaining == 0;
        $ready.wait;
    }
}
```
One could now write
```
$l.protect: {
    $ready.wait: { $remaining == 0 }
}
```

### CODE REVIEW

File: src/core.c/Lock.pm6 [hunk 1]
```diff
@@ -9,7 +9,9 @@ my class Lock {
         method new() {
             X::Lock::ConditionVariable::New.new.throw
         }
-        method wait() { nqp::condwait(self) }
+        proto method wait(|) {*}
+        multi method wait() { nqp::condwait(self) }
+        multi method wait(&predicate) { nqp::condwait(self) while !predicate; }
```
Comment (vrurg) [inline src/core.c/Lock.pm6:14]: Why not `until predicate`?

Comment (Leont) [reply src/core.c/Lock.pm6:14]: Because I only thought of that after I wrote the code?

**Comment by vrurg** [comment]
One thing I'd like to ask you about is to be somewhat more elaborate in commit messages. It's so helpful for code archeologists. :)

Will merge when get back home, unless somebody does it by then.

**Comment by Leont** [comment]
> One thing I'd like to ask you about is to be somewhat more elaborate in commit messages. It's so helpful for code archeologists. :)

Fair point!

File: src/core.c/Lock.pm6 [hunk 2]
```diff
@@ -9,7 +9,9 @@ my class Lock {
         method new() {
             X::Lock::ConditionVariable::New.new.throw
         }
-        method wait() { nqp::condwait(self) }
+        proto method wait(|) {*}
+        multi method wait() { nqp::condwait(self) }
+        multi method wait(&predicate) { nqp::condwait(self) until predicate; }
```
Comment (jnthn) [inline src/core.c/Lock.pm6:14]: Hm, I doubt these return anything useful or that we've spec'd that, but I'd suggest while we're at this, we throw a `--> Nil` on the two candidates.

**Review by jnthn — COMMENTED** [review]
This makes sense to me; it's very much in the spirit of providing factored out forms of common use cases (aside from the `protect` method on locks, the same kind of thing is done by proving a block form of `cas` too).

File: src/core.c/Lock.pm6 (hunk 2 above)
Comment (Leont) [reply src/core.c/Lock.pm6]: Done

**Review by jnthn — APPROVED** [review]

**Pull request merged**